salt/states/boto_ec2.py

```diff
@@ -452,3 +452,151 @@ def snapshot_created(name, ami_name, instance_name, wait_until_available=True, w
         sleep(5)
 
     return ret
+
+
+def instance_present(name, instance_name=None, instance_id=None, image_id=None,
+                     image_name=None, tags=None, key_name=None,
+                     security_groups=None, user_data=None, instance_type=None,
+                     placement=None, kernel_id=None, ramdisk_id=None,
+                     vpc_id=None, vpc_name=None, monitoring_enabled=None,
+                     subnet_id=None, subnet_name=None, private_ip_address=None,
+                     block_device_map=None, disable_api_termination=None,
+                     instance_initiated_shutdown_behavior=None,
+                     placement_group=None, client_token=None,
+                     security_group_ids=None, security_group_names=None,
+                     additional_info=None, tenancy=None,
+                     instance_profile_arn=None, instance_profile_name=None,
+                     ebs_optimized=None, network_interfaces=None,
+                     attributes=None, target_state=None, region=None, key=None,
+                     keyid=None, profile=None):
+    ### TODO - implement 'target_state'...
+    '''
+    Ensure an EC2 instance is running with the given attributes and state.
+    '''
+    ret = {'name': name,
+           'result': True,
+           'comment': '',
+           'changes': {}
+          }
+    _create = False
+    running_states = ('pending', 'rebooting', 'running', 'stopping', 'stopped')
+
+    if not instance_id:
+        try:
+            instance_id = __salt__['boto_ec2.get_id'](name=instance_name if instance_name else name,
+                                                      tags=tags, region=region, key=key, keyid=keyid,
+                                                      profile=profile, in_states=running_states)
+        except CommandExecutionError as e:
+            ret['result'] = None
+            ret['comment'] = 'Couldn\'t determine current status of instance {0}.'.format(instance_name)
+            return ret
+
+    exists = __salt__['boto_ec2.exists'](instance_id=instance_id, region=region,
+                                         key=key, keyid=keyid, profile=profile)
+    if not exists:
+        _create = True
+    else:
+        instances = __salt__['boto_ec2.find_instances'](instance_id=instance_id, region=region,
+                                                        key=key, keyid=keyid, profile=profile,
+                                                        return_objs=True, in_states=running_states)
+        if not len(instances):
+            _create = True
+
+    if _create:
+        r = __salt__['boto_ec2.run'](image_id, instance_name if instance_name else name,
```

Comment:
Missing test mode check. This will create instances when people call the state with test=true